src/backend/optimizer/plan/createplan.c

```diff
@@ -2026,8 +2028,9 @@ create_projection_plan(PlannerInfo *root, ProjectionPath *best_path, int flags)
 		 * Tell create_plan_recurse that we're going to ignore the tlist it
 		 * produces.
 		 */
-		subplan = create_plan_recurse(root, best_path->subpath,
-									  CP_IGNORE_TLIST);
+		bool is_initplan = is_function_scan_on_initplan(root, best_path->subpath);
+		int flags = is_initplan ? 0 : CP_IGNORE_TLIST;
+		subplan = create_plan_recurse(root, best_path->subpath, flags);
 		Assert(is_projection_capable_plan(subplan));
 		tlist = build_path_tlist(root, &best_path->path);
```

Comment:
As far as I know, yes. The initplan's target list is the one with the default ordering (as returned by the function itself), so it should be stored into tuplestore in that order. The function scan then reads that tuplestore with the actual target list, reordering and modifying the attributes as needed. This should happen only once during the read of the tuplestore, and the test covers exactly this case anyway. I'll still research this a bit more and find the exact place where tuplestore applies the projection.

Reply:
So, the tuplestore is created and filled in [ExecMakeTableFunctionResult](https://github.com/arenadata/gpdb/blob/9bfb1166292470f16612ce6f34d48f764a784700/src/backend/executor/execSRF.c#L101). The tuplestore is then read in [FunctionNext_guts](https://github.com/arenadata/gpdb/blob/9bfb1166292470f16612ce6f34d48f764a784700/src/backend/executor/nodeFunctionscan.c#L113). The projection is then happening in [ExecScan](https://github.com/arenadata/gpdb/blob/9bfb1166292470f16612ce6f34d48f764a784700/src/backend/executor/execScan.c#L245), much after tuplestore reading. This means that if the tuplestore is stored in the same ordering as the original function returns values, it will be projected the same way as if the tuplestore wasn't there (as we expect). This ordering is called "physical target list" in code, and decision to construct it is made with [use_physical_tlist](https://github.com/arenadata/gpdb/blob/9bfb1166292470f16612ce6f34d48f764a784700/src/backend/optimizer/plan/createplan.c#L875). It checks the following conditions:

1. flags `CP_EXACT_TLIST` and `CP_SMALL_TLIST` are not set (true, flags are 0 in this case).
2. `rtekind` is one of the supported ones (true, `RTE_FUNCTION` is supported).
3. `reloptkind` must be `RELOPT_BASEREL` (true).
4. path is not `CustomPath` (true, it is `Path`)
5. additional condition for `BitmapHeapPath` (irrelevant, our path is `FunctionScan`)
6. system columns are not requested (I doubt a function can return system columns without projecting them to normal values)
7. no placeholder expressions (Functions can't return that)
8. additional condition for `IndexOnlyScan` (irrelevant, our path is `FunctionScan`)
9. additional condition if flag `CP_LABEL_TLIST` is set (irrelevant, flags are 0)
10. table is not append-optimized (it isn't).

Therefore, physical target list will always be constructed and so the ordering in the tuplestore will always match the ordering returned by the function. Projection is applied only once after the tuplestore read, and is [skipped](https://github.com/arenadata/gpdb/blob/9bfb1166292470f16612ce6f34d48f764a784700/src/backend/executor/execScan.c#L189) during the write, so this projection should always be correct.

Reply:
Thanks for detailed investigation! Great job